File: Swift/프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.swift
func solution(_ dartResult:String) -> Int {
    var scores: [Int] = []
    var currentNumber = ""
    
    let dartArray = Array(dartResult)
    let numSet: Set<Character> = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"]
    
    for i in 0..<dartArray.count {
        let char = dartArray[i]
        
        if numSet.contains(char) {
            currentNumber.append(char)
        } else if char == "S" || char == "D" || char == "T" {
            guard let score = Int(currentNumber) else { continue }
            
            if char == "S" {
                scores.append(score)
            } else if char == "D" {
                scores.append(score * score)
            } else if char == "T" {
                scores.append(score * score * score)
            }
            currentNumber = ""
        } else if char == "*" {
            let count = scores.count
            if count >= 1 {
                scores[count - 1] *= 2
            }
            if count >= 2 {
                scores[count - 2] *= 2
            }
        } else if char == "#" {
            let count = scores.count
            if count >= 1 {
                scores[count - 1] *= -1
            }
        }
    }
    
    var answer = 0
    for score in scores {
        answer += score
    }
    return answer
}

File: Java/프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.ArrayList;

class Solution {
    public int solution(String dartResult) {
        int length = dartResult.length();
        ArrayList<Integer> scores = new ArrayList<>();
        char[] dartArray = dartResult.toCharArray();
        
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < length; i++) {
            char c = dartArray[i];
            
            if (c >= '0' && c <= '9') {
                sb.append(c);
            } else if (c == 'S' || c == 'D' || c == 'T') {
                int score = Integer.parseInt(sb.toString());
                
                if (c == 'S') {
                    scores.add(score);
                } else if (c == 'D') {
                    scores.add(score * score);
                } else if (c == 'T') {
                    scores.add(score * score * score);
                }
                
                sb.setLength(0);
            } else if (c == '*') {
                int sLength = scores.size();
                if (sLength >= 1) {
                    scores.set(sLength - 1, scores.get(sLength - 1) * 2);
                }
                if (sLength >= 2) {
                    scores.set(sLength - 2, scores.get(sLength - 2) * 2);
                }
            } else if (c == '#') {
                int sLength = scores.size();
                if (sLength >= 1) {
                    scores.set(sLength - 1, scores.get(sLength - 1) * -1);
                }
            }
        }
        
        int answer = 0;
        for (int score : scores) {
            answer += score;
        }
        return answer;
    }
}